addons/stock/models/stock_inventory.py

```diff
@@ -200,46 +200,40 @@ def action_view_related_move_lines(self):
 
     def _get_inventory_lines_values(self):
         # TDE CLEANME: is sql really necessary ? I don't think so
-        locations = self.env['stock.location']
+        vals = []
+        domain = [('quantity', '!=', 0.0), ('product_id.active', '=', True)]
         if self.location_ids:
-            locations = self.env['stock.location'].search([('id', 'child_of', self.location_ids.ids)])
+            locations_ids = self.env['stock.location'].search_read([('id', 'child_of', self.location_ids.ids)], ['id'])
         else:
-            locations = self.env['stock.location'].search(self._domain_location_ids())
-        domain = ' location_id in %s AND quantity != 0 AND active = TRUE'
-        args = (tuple(locations.ids),)
-
-        vals = []
-        Product = self.env['product.product']
-        # Empty recordset of products available in stock_quants
-        quant_products = self.env['product.product']
+            locations_ids = self.env['stock.location'].search_read(self._domain_location_ids(), ['id'])
+        locations_ids = [l['id'] for l in locations_ids]
+        domain = expression.AND([domain, [('location_id', 'in', locations_ids)]])
 
-        # If inventory by company
         if self.company_id:
-            domain += ' AND company_id = %s'
-            args += (self.company_id.id,)
+            domain = expression.AND([domain, [('company_id', '=', self.company_id.id)]])
         if self.product_ids:
-            domain += ' AND product_id in %s'
-            args += (tuple(self.product_ids.ids),)
-
-        self.env.cr.execute("""SELECT product_id, sum(quantity) as product_qty, location_id, lot_id as prod_lot_id, package_id, owner_id as partner_id
-            FROM stock_quant
-            LEFT JOIN product_product
-            ON product_product.id = stock_quant.product_id
-            WHERE %s
-            GROUP BY product_id, location_id, lot_id, package_id, partner_id """ % domain, args)
-
-        for product_data in self.env.cr.dictfetchall():
-            product_data['inventory_id'] = self.id
-            # replace the None the dictionary by False, because falsy values are tested later on
-            for void_field in [item[0] for item in product_data.items() if item[1] is None]:
-                product_data[void_field] = False
-            product_data['theoretical_qty'] = product_data['product_qty']
+            domain = expression.AND([domain, [('product_id', 'in', self.product_ids.ids)]])
+
+        fields = ['product_id', 'quantity:sum', 'location_id', 'lot_id', 'package_id', 'owner_id']
+        group_by = ['product_id', 'location_id', 'lot_id', 'package_id', 'owner_id']
+
+        quants_groups = self.env['stock.quant'].read_group(domain, fields, group_by, lazy=False)
```

Comment:
If `load=''` read parameter were supported then you didn't need re-transform the dictionary.

Maybe another change for master:
 - https://github.com/odoo/odoo/issues/34432